README.md
# Zi

Zi is a maven plugin for clojure. It does something similar to
[clojure-maven-plugin](http://github.com/talios/clojure-maven-plugin), but does
so differently.

It uses the maven pom `sourceDirectory` and `testSourceDirectory` settings to
locate source, which by default means that it uses `src/main/clojure` and
`src/test/clojure`. The goals should work with a `pom.xml` generated by
[leiningen](https://github.com/technomancy/leiningen).

From an implementation perspecitve, most of the goals are written in clojure.

Zi requires maven 3.0.4.

## Available goals

 * zi:resources
 * zi:testResources
 * zi:compile
 * zi:ritz
 * zi:ritz-nrepl
 * zi:swank-clojure
 * zi:test
 * zi:codox
 * zi:marginalia
 * zi:ring-genfiles

## Install

Globally installing the plugin allows you to run the goals without modifying
a project's pom file.

To globally enable the zi plugin, you need to add `pluginGroup` and
`pluginRepository` configuration to your `~/.m2/settings.xml` file.

```xml
    <pluginGroups>
      <pluginGroup>org.cloudhoist.plugin</pluginGroup>
    </pluginGroups>
```

```xml
    <profiles>
      <profile>
        <id>clojure-dev</id>
        <pluginRepositories>
          <pluginRepository>
            <id>sonatype-snapshots</id>
            <url>http://oss.sonatype.org/content/repositories/releases</url>
          </pluginRepository>
        </pluginRepositories>
      </profile>
    </profiles>

    <activeProfiles>
      <activeProfile>clojure-dev</activeProfile>
    </activeProfiles>
```

To enable zi in a project pom, without globally enabling the plugin, you will
need to add a `pluginRepositories` entry to your pom:

```xml
    <pluginRepositories>
      <pluginRepository>
        <id>sonatype-snapshots</id>
        <url>http://oss.sonatype.org/content/repositories/releases</url>
      </pluginRepository>
    </pluginRepositories>
```

## Configuration

It uses the maven pom `sourceDirectory` and `testSourceDirectory` settings to
locate source, which by default means that it uses `src/main/clojure` and
`src/test/clojure`.

## Goals

### resources

The resources goal copies clojure source to the target. This is probably what
you need to make sure that your clj source files end up in your jar file.

```xml
    <build>
      <plugins>
        <plugin>
          <groupId>org.cloudhoist.plugin</groupId>
          <artifactId>zi</artifactId>
          <version>0.5.5</version>
          <executions>
            <execution>
              <id>default-resources</id>
              <phase>process-resources</phase>
              <goals>
                <goal>resources</goal>
              </goals>
            </execution>
          </executions>
        </plugin>
      </plugins>
    </build>
```

### testResources

The testResources goal copies clojure test source to the target. This is
probably what you need to make sure that your clj source files end up in your
test-jar file.

```xml
    <build>
      <plugins>
        <plugin>
          <groupId>org.cloudhoist.plugin</groupId>
          <artifactId>zi</artifactId>
          <version>0.5.5</version>
          <executions>
            <execution>
              <id>default-test-resources</id>
              <phase>process-test-resources</phase>
              <goals>
                <goal>test-resources</goal>
              </goals>
            </execution>
          </executions>
        </plugin>
      </plugins>
    </build>
```

### compile

The compile goal AOT compiles clojure source.

```xml
    <build>
      <plugins>
        <plugin>
          <groupId>org.cloudhoist.plugin</groupId>
          <artifactId>zi</artifactId>
          <version>0.5.5</version>
          <executions>
            <execution>
              <id>default-compile</id>
              <goals>
                <goal>compile</goal>
              </goals>
            </execution>
          </executions>
          <configuration>
            <excludes>
              <exclude>**/test.clj</exclude>
            </excludes>
          </configuration>
        </plugin>
      </plugins>
    </build>
```

<table>
  <tr>
    <th>Property</th>
    <th>Variable</th>
    <th>Default</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>includes</td>
    <td>
    <td>**/*.clj</td>
    <td>A set of source patterns to include</td>
  </tr>
  <tr>
    <td>excludes</td>
    <td>
    <td></td>
    <td>A set of source patterns to exclude</td>
  </tr>
</table>

### ritz

The ritz goal starts a ritz server.

<table>
  <tr>
    <th>Property</th>
    <th>Variable</th>
    <th>Default</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>port</td>
    <td>clojure.swank.port</td>
    <td>4005</td>
    <td>The swank server port</td>
  </tr>
  <tr>
    <td>encoding</td>
    <td>clojure.swank.encoding</td>
    <td>iso-8859-1</td>
    <td>The swank encoding to use</td>
  </tr>
</table>

The ritz, ritz-nrepl and swank-clojure goals all recognise sub-projects linked
in a checkouts directory, and adds their sources and resources to the the
classpath.

The ritz and ritz-nrepl goals also adds source jars to the classpath if they are
available in the local repository.  You can resolve the source jars using `mvn
dependency:sources`.

### ritz-nrepl

The ritz-nrepl goal starts a ritz nREPL server.

<table>
  <tr>
    <th>Property</th>
    <th>Variable</th>
    <th>Default</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>port</td>
    <td>clojure.nrepl.port</td>
    <td>4005</td>
    <td>The nrepl server port</td>
  </tr>
</table>

### swank-clojure

The swank-clojure goal starts a swank-clojure server.

<table>
  <tr>
    <th>Property</th>
    <th>Variable</th>
    <th>Default</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>port</td>
    <td>clojure.swank.port</td>
    <td>4005</td>
    <td>The swank server port</td>
  </tr>
  <tr>
    <td>encoding</td>
    <td>clojure.swank.encoding</td>
    <td>iso-8859-1</td>
    <td>The swank encoding to use</td>
  </tr>
</table>

### test

The test goal runs clojure.test tests.

<table>
  <tr>
    <th>Property</th>
    <th>Variable</th>
    <th>Default</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>initScript</td>
    <td>clojure.initScript</td>
    <td></td>
    <td>A clojure source string that is run before the tests</td>
  </tr>
</table>

### codox

The codox goal creates codox generated API documentation in the `doc` directory.

<table>
  <tr>
    <th>Property</th>
    <th>Variable</th>
    <th>Default</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>codoxTargetDirectory</td>
    <td></td>
    <td>${project.build.directory}/doc</td>
    <td>The directory where codox should write its output</td>
  </tr>
</table>

### marginalia

The marginalia goal creates a marginalia annotated source page.

<table>
  <tr>
    <th>Property</th>
    <th>Variable</th>
    <th>Default</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>marginaliaTargetDirectory</td>
    <td></td>
    <td>${project.build.directory}</td>
    <td>The directory where marginalia should write uberdoc.html</td>
  </tr>
</table>

### ring-genfiles

The ring-genfiles goal creates files required for a war packaging.
See [lein-ring plugin](https://github.com/weavejester/lein-ring/).

<table>
  <tr>
    <th>Property</th>
    <th>Default</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>ringServletClass</td>
    <td></td>
    <td>The name of the servlet class</td>
  </tr>
  <tr>
    <td>ringServletName</td>
    <td></td>
    <td>The name of the servlet</td>
  </tr>
  <tr>
    <td>ringListenerClass</td>
    <td></td>
    <td>The name of the ServletContextListener class</td>
  </tr>
  <tr>
    <td>ringHandler</td>
    <td></td>
    <td>The name of the ring handler</td>
  </tr>
  <tr>
    <td>ringInit</td>
    <td></td>
    <td>The name of an initialisation hook function</td>
  </tr>
  <tr>
    <td>ringDestroy</td>
    <td></td>
    <td>The name of a shutdown hook function</td>
  </tr>
  <tr>
    <td>ringUrlPattern</td>
    <td></td>
    <td>The url pattern of the servlet mapping (in web.xml). Defaults to "/*".</td>
  </tr>
</table>

## Zi, the builder

Zi was a builder in
[northern mythology](http://www.pitt.edu/~dash/mbuilder.html#eckwadt).

## License

Licensed under [EPL](http://www.eclipse.org/legal/epl-v10.html)

Copyright 2011 Hugo Duncan.


ReleaseNotes.md
# Release Notes

# 0.5.5

- Add repository information for clojars
  When using zi goals, clojars artifacts were not being resolved. Fixes #11

# 0.5.4

- Add nrepl goal
  The nrepl goal starts a vanilla nREPL server.

- Add a ritz-nrepl goal
  Adds a goal for starting a ritz-nrepl server.

# 0.5.3

- Respect maven.test.skip in the test task

- Update to ritz 0.4.1. Now works with ritz 0.3.x and 0.4.x.

# 0.5.2

- Update to clojure-maven 0.3.3 to fix line number reporting in compiler.
  Fixes #10

# 0.5.1

- in ring-genfiles, fix handler-source which was not returning wished fn

# 0.5.0

- Properly handle classpath for targets running in project classpath
  The dependencies for codox, marginalia, ritz, and swank-clojure are now
  handled properly.  The zi pom no longer depends on these projects.

- Automatically add tools.jar to classpath for the ritz task

- Use leiningen-core for calculating checkout dependency paths
  Upgrade to clojure 1.3.0 and leiningen-core 2.0.0-preview4

- Add clojure-maven-mojo as a dependency, and use it's log wrapper

- Add .ritz-exception-filters to .gitignore

- Add a test for the test mojo runner

- Update to marginalia 0.7.0, and run it in the project classpath

- Remove use of contrib in ring-genfiles
  Switch to data.xml

- Unindent pom description for marginalia and codox

- Update to maven 3.0.4 and aether 1.13

# 0.4.5

- Allow exclusion of tests based on a regex
  Provides a mechanism for disabling clojure.test test namespaces

- Allow specification of codox output writer

- Allow explicit specification of codox api version

# 0.4.4
- Add support for codoxTargetDirectory

- Add a codox goal
  The codox goal can be used to generate api documentation based on the
  codox project

- Allow setting of ritz loglevel via a system property
  Setting ritz.logglevel=:trace will enable ritz logging

## 0.4.3
- Allow tests to override core.test/report before redef'ing.
  Some test frameworks, such as midje register additional defmethods for
  core.test/report. By requiring the test namespaces before the report var
  is redef'ed to a function, these test frameworks become supported.

- Add note about source jars to readme

## 0.4.2

- ring gen files : listener-class and servlet-class in web.xml should point
  to the class file (no '-', use '_' instead)

- Update to ritz 0.2.0

- Enforce maven 3.0.3 requirement

## 0.4.1

- Update initScript handling in test mojo to handle multiple forms
  Only the first form was being evaluated.

## 0.4.0

Adds the ring-genfiles goal for lein-ring compatible Servlet,
ServletContextListener and web.xml generation.

- Fix an issue with keywords used (init and destroy)

- ring-genfiles formatting, add to readme, and add integration test

- Add ring-genfiles goal to allow servlet/listener/web.xml generation

## 0.3.10

- Fix for non-string used as a test message crashing zi:test

## 0.3.9

- Fix an issue preventing the test task from working in clojure-1.3.0
  The form passed for portable-redef was being quoted in clojure-1.2, and
  run in 1.3. The quoting was adding zi.test to the with-redefs symbol.

## 0.3.8

- Update for classlojure 0.6.1

## 0.3.7

- Remove class files from final jar
  The mojo extractor was creating class files that were packaged.

- Add initScript property for test target

## 0.3.6

- Fix skipTests handling in test mojo

### Fixes

## 0.3.5

### Features

- Honour skipTests, and fix for case when no tests defined

- Add available source jars to classpath

### Fixes
- Remove superfluous log statement

- Add sonatype repository to pom


## 0.3.4

- Throw a MojoFailureException when tests fail

- Don't run ritz on pom packaged projects

- Add testResources goal

## 0.3.3

- Ensure tests run inside a clojure.main/with-bindings
  Code that calls set! on common rebindable vars (such as
  *warn-on-reflection*) fails otherwise

- Fix the classpath elements used in ritz and test

## 0.3.2

- Fix resources mojo to write to target/classes

## 0.3.1

- Really add marginalia goal

- Add integration testing

## 0.3.0

- Add marginalia goal

- Factor out overridable-artifact-path

## 0.2.2

- Fix mojo/defmojo to actually read the project source paths

## 0.2.1

- Add checkouts to ritz and swank-clojure goals

- Update readme with new goals


## 0.2.0

- Allow overriding of ritz and swank-clojure artifacts

- Add a test goal for running clojure.test tests

- Add swank-clojure goal

- Update ritz mojo to use defmojo

- Add resources goal for copying clojure source

- Add macro for better mojo definition syntax

- Add compiler mojo
  The compile goal compiles clojure source. It respects the options from
  maven-compiler-plugin.

## 0.1.0

- Initial release


src/main/clojure/zi/ritz.clj
(ns zi.ritz
  "ritz mojo for zi plugin"
  (:require
   [zi.core :as core]
   [zi.mojo :as mojo]
   [clojure.maven.mojo.log :as log]
   [zi.checkouts :as checkouts])
  (:use
   [zi.maven :only [resolve-dependency]])
  (:import
   java.io.File
   [clojure.maven.annotations
    Goal RequiresDependencyResolution Parameter Component]
    org.apache.maven.plugin.ContextEnabled
    org.apache.maven.plugin.Mojo
    org.apache.maven.plugin.MojoExecutionException))

(mojo/defmojo RitzMojo
  {Goal "ritz"
   RequiresDependencyResolution "test"}
  [^{Component {:role "org.sonatype.aether.RepositorySystem"
                :alias "repoSystem"}}
   repo-system

   ^{Component {:role "org.apache.maven.project.ProjectBuilder"
                :alias "projectBuilder"}}
   project-builder

   ^{Parameter {:defaultValue "${repositorySystemSession}" :readonly true
                :alias "repoSystemSession"}}
   repo-system-session

   ^{Parameter
     {:expression "${clojure.swank.port}" :defaultValue "4005"}}
   ^Integer
   port

   ^{Parameter
     {:expression "${clojure.swank.encoding}" :defaultValue "iso-8859-1"}}
   ^String
   encoding

   ^{Parameter
     {:expression "${clojure.swank.network}" :defaultValue "localhost"}}
   ^String
   network

   ^{Parameter
     {:defaultValue "${project.packaging}" :required true}}
   ^String
   packaging

   ^{Parameter
     {:expression "${project}"
      :description "Project"}}
   project]

  (if (= packaging "pom")
    (log/info "Ritz can not be run on a project with pom packaging")
    (let [tmpfile (try
                    (File/createTempFile "swank" ".port")
                    (catch java.io.IOException e
                      (throw
                       (MojoExecutionException.
                        "Could not create swank port file" e))))
          ritz-version (or (System/getProperty "ritz.version") "0.4.1")
          ritz-0-3 (.startsWith ritz-version "0.3")
          ritz-artifacts (resolve-dependency
                         repo-system
                         repo-system-session
                         (.getRemoteProjectRepositories project)
                         "ritz" (if ritz-0-3 "ritz" "ritz-swank")
                         ritz-version
                         {})
          source-paths (->
                        (core/clojure-source-paths source-directory)
                        (into (core/clojure-source-paths test-source-directory))
                        (into (checkouts/checkout-paths
                               repo-system repo-system-session
                               project-builder)))
          classpath-elements (concat test-classpath-elements ritz-artifacts)
          log-level (if-let [p (System/getProperty "ritz.loglevel")]
                      (read-string p)
                      :warn)]
      (log/debugf "source paths: %s" (vec source-paths))
      (log/debugf "classpath elements: %s" (vec classpath-elements))
      (log/debugf "log-level %s" (pr-str log-level))
      (core/eval-clojure
       source-paths
       (-> classpath-elements
           core/classpath-with-source-jars
           core/classpath-with-tools-jar)
       `(do
          (require '~(if ritz-0-3
                       'ritz.socket-server
                       'ritz.swank.socket-server))
          (~(if ritz-0-3
              'ritz.socket-server/start
              'ritz.swank.socket-server/start)
           {:port-file ~(.getPath tmpfile)
            :host ~network
            :port ~(Integer/parseInt port)
            :encoding ~encoding
            :dont-close true
            :log-level ~log-level}))))))
